vyper/semantics/analysis/data_positions.py

```diff
@@ -244,73 +268,117 @@ def _allocate_nonreentrant_keys(vyper_module, allocators):
         size = varinfo.get_size()
 
         # CMC 2021-07-23 note that HashMaps get assigned a slot here
         # using the same allocator (even though there is not really
         # any risk of physical overlap)
         offset = allocator.allocate_slot(size, node.target.id, node)
-
         varinfo.set_position(VarOffset(offset))
 
+
+# get the layout for export
+def generate_layout_export(vyper_module: vy_ast.Module):
+    return _generate_layout_export_r(vyper_module)
+
+
+def _generate_layout_export_r(vyper_module):
+    ret: defaultdict[str, InsertableOnceDict[str, dict]] = defaultdict(InsertableOnceDict)
+
+    for node in _get_allocatable(vyper_module):
+        if isinstance(node, vy_ast.InitializesDecl):
+            module_info = node._metadata["initializes_info"].module_info
+            module_layout = _generate_layout_export_r(module_info.module_node)
+            module_alias = module_info.alias
+            for layout_key in module_layout.keys():
+                assert layout_key in _LAYOUT_KEYS.values()
+
+                # lift the nonreentrancy key (if any) into the outer dict
+                # note that lifting can leave the inner dict empty, which
+                # should be filtered (below) for cleanliness
+                nonreentrant = module_layout[layout_key].pop(GLOBAL_NONREENTRANT_KEY, None)
+                if nonreentrant is not None and GLOBAL_NONREENTRANT_KEY not in ret[layout_key]:
+                    ret[layout_key][GLOBAL_NONREENTRANT_KEY] = nonreentrant
+
+                # add the module as a nested dict, but only if it is non-empty
+                if len(module_layout[layout_key]) != 0:
+                    ret[layout_key][module_alias] = module_layout[layout_key]
+
+            continue
+
+        assert isinstance(node, vy_ast.VariableDecl)
+        varinfo = node.target._metadata["varinfo"]
+        # skip non-state variables
+        if not varinfo.is_state_variable():
+            continue
+
+        location = varinfo.location
         layout_key = _LAYOUT_KEYS[location]
         type_ = varinfo.typ
+        size = varinfo.get_size()
+        offset = varinfo.position.position
+
         # this could have better typing but leave it untyped until
         # we understand the use case better
         if location == DataLocation.CODE:
             item = {"type": str(type_), "length": size, "offset": offset}
         elif location in (DataLocation.STORAGE, DataLocation.TRANSIENT):
-            item = {"type": str(type_), "slot": offset}
+            item = {"type": str(type_), "n_slots": size, "slot": offset}
         else:  # pragma: nocover
             raise CompilerPanic("unreachable")
         ret[layout_key][node.target.id] = item
 
+    for fn in vyper_module.get_children(vy_ast.FunctionDef):
+        fn_t = fn._metadata["func_type"]
+        if not fn_t.nonreentrant:
+            continue
+
+        location = get_reentrancy_key_location()
```

Comment:
with overrides, non-reentrant key will be in storage, but the layout (`-f`) will return that key in the `transient_storage_layout`

Reply:
right, so you're saying the output of `-f layout` won't match the override file right?

Reply:
thats right

Reply:
b9acf9829e1f61155b1af57a64484e76e01ff8d2, do you think this addresses it appropriately?